### PR TITLE
[Semi-Auto] Replace the pointers in reshape-like spmd rule with smart pointers.

### PR DESCRIPTION
**Pull request by pkuzyc**
### PR types
Others

### PR changes
Others

### Description
 Pcard-76459
Replace the pointers in reshape-like spmd rule with smart pointers.

### CODE REVIEW

File: paddle/phi/infermeta/spmd_rules/dim_trans.cc [hunk 1]
```diff
@@ -93,27 +83,26 @@ std::string Flatten::to_string() {
...
-}
+Split::~Split() { std::vector<int64_t>().swap(splitted_shape_); }
 
-DimTrans* Split::input() const { return input_dim_trans_; }
+const std::shared_ptr<DimTrans>& Split::input() const {
+  return input_dim_trans_;
+}
 
-void Split::set_input(DimTrans* dim) { input_dim_trans_ = dim; }
+void Split::set_input(const std::shared_ptr<DimTrans>& dim) {
```
Comment (deepllz) [inline paddle/phi/infermeta/spmd_rules/dim_trans.cc]: shared_ptr类型的参数，是否有必要使用引用，一般建议使用值传参即可。

File: paddle/phi/infermeta/spmd_rules/dim_trans.cc [hunk 2]
```diff
@@ -166,55 +167,48 @@ DimTrans* make_split(DimTrans* dim,
...
-DimTrans* GetDimTrans(DimTrans* dim_trans,
-                      std::vector<std::vector<bool>>* shardable,
-                      std::set<int64_t>* seen_dims,
-                      const std::vector<int64_t>& input_shape,
-                      const std::vector<int64_t>& mesh_shape,
-                      const std::vector<int64_t>& input_dims_mapping,
-                      const std::set<int64_t>& sharded_input_dims) {
+std::shared_ptr<DimTrans> GetDimTrans(
+    const std::shared_ptr<DimTrans>& dim_trans,
+    std::vector<std::vector<bool>>* shardable,
```
Comment (deepllz) [inline paddle/phi/infermeta/spmd_rules/dim_trans.cc]: 参数顺序：这里是否考虑把输出参数放到输入参数的后面？

File: paddle/phi/infermeta/spmd_rules/dim_trans.cc (hunk 1 above)
Comment (pkuzyc) [reply paddle/phi/infermeta/spmd_rules/dim_trans.cc]: Done

File: paddle/phi/infermeta/spmd_rules/dim_trans.cc (hunk 2 above)
Comment (pkuzyc) [reply paddle/phi/infermeta/spmd_rules/dim_trans.cc]: Done

**Review by deepllz — APPROVED** [review]
LGTM

**Review by zhiqiu — APPROVED** [review]
LGTM

**Pull request merged**